crates/pathfinder/src/sync/catch_up.rs

```diff
@@ -0,0 +1,1706 @@
+            } = class;
+
+            match definition {
+                class_definitions::CompiledClassDefinition::Cairo(cairo) => db_tx
+                    .update_cairo_class(hash, &cairo)
+                    .context("Inserting cairo class definition"),
+                crate::sync::class_definitions::CompiledClassDefinition::Sierra {
+                    sierra_definition,
+                    casm_definition,
+                } => {
+                    let sierra_hash = SierraHash(hash.0);
+                    let casm_hash = db_tx
+                        .casm_hash(hash)
+                        .context("Getting casm hash")?
+                        .context("Casm not found")?;
+                    db_tx
+                        .update_sierra_class(
+                            &sierra_hash,
+                            &sierra_definition,
+                            &casm_hash,
+                            &casm_definition,
+                        )
+                        .context("Inserting sierra class definition")
+                }
+            }
+        })?;
+
+        let result = db_tx
+            .commit()
+            .context("Committing transaction")
+            .map(|_| (block_number, header.hash));
+
+        tracing::debug!(number=%block_number, "Block stored");
+
+        result
+    }
+}
+
+pub mod gateway {
+    use std::future::Future;
+    use std::sync::Arc;
+
+    use anyhow::Context;
+    use pathfinder_common::{
+        BlockCommitmentSignature,
+        BlockHash,
+        BlockHeader,
+        BlockId,
+        BlockNumber,
+        CasmHash,
+        ClassHash,
+        ConsensusInfo,
+        EventCommitment,
+        GasPrice,
+        ReceiptCommitment,
+        SequencerAddress,
+        SierraHash,
+        StateCommitment,
+        StateDiffCommitment,
+        StateUpdate,
+        TransactionCommitment,
+    };
+    use pathfinder_crypto::Felt;
+    use pathfinder_ethereum::{EthereumApi, EthereumStateUpdate};
+    use pathfinder_merkle_tree::starknet_state::update_starknet_state;
+    use pathfinder_rpc::tracker::SubmittedTransactionTracker;
+    use pathfinder_storage::{Connection, Storage, Transaction, TransactionBehavior};
+    use starknet_gateway_client::GatewayApi;
+    use starknet_gateway_types::reply::{
+        Block,
+        GasPrices,
+        PendingBlock,
+        PreConfirmedBlock,
+        PreLatestBlock,
+    };
+    use tokio::sync::{mpsc, watch};
+
+    use crate::state::l1::L1SyncContext;
+    use crate::state::l2::{self, BlockChain, L2SyncContext};
+    use crate::state::{SyncContext, SyncEvent};
+
+    pub fn spawn<Ethereum, SequencerClient, F1, F2, L1Sync, L2CatchUp>(
+        context: SyncContext<SequencerClient, Ethereum>,
+        mut catch_up_start: watch::Receiver<Option<u64>>,
+        consensus_info_rx: watch::Receiver<Option<ConsensusInfo>>,
+        sync_event_tx: mpsc::Sender<SyncEvent>,
+        // TODO: Do we do L1 sync here?
```

Comment:
I think we should.

L1 sync at the moment syncs the latest L1 confirmed block number/hash only, which is required for us to be able to return correct block status via JSON-RPC.